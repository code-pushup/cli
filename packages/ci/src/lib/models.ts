import type { SourceFileIssue } from './issues';
import type { MonorepoTool } from './monorepo';

/**
 * Customization options for {@link runInCI}
 * @see https://github.com/code-pushup/cli/tree/main/packages/ci#options
 */
export type Options = {
  monorepo?: boolean | MonorepoTool;
  projects?: string[] | null;
  task?: string;
  bin?: string;
  config?: string | null;
  directory?: string;
  silent?: boolean;
  debug?: boolean;
  detectNewIssues?: boolean;
  logger?: Logger;
};

/**
 * {@link Options} with filled-in defaults.
 */
export type Settings = Required<Options>;

/**
 * Branches/commits for {@link runInCI}
 */
export type GitRefs = {
  head: GitBranch;
  base?: GitBranch;
};

/**
 * API client for {@link runInCI} - adapter for CI provider
 * @see https://github.com/code-pushup/cli/tree/main/packages/ci#provider-api-client
 */
export type ProviderAPIClient = {
  maxCommentChars: number;
  downloadReportArtifact?: () => Promise<string | null>;
  listComments: () => Promise<Comment[]>;
  updateComment: (id: number, body: string) => Promise<Comment>;
  createComment: (body: string) => Promise<Comment>;
};

/**
 * PR comment from {@link ProviderAPIClient}
 */
export type Comment = {
  id: number;
  body: string;
  url: string;
};

/**
 * Git ref (e.g. 'main') and commit sha (e.g. '9d7568265bd5b17d1bb51a6eb6407e0d387058d2')
 */
export type GitBranch = {
  ref: string;
  sha: string;
};

/**
 * Logger instance (e.g. `console`) for reporting progress and problems
 */
export type Logger = {
  error: (message: string) => void;
  warn: (message: string) => void;
  info: (message: string) => void;
  debug: (message: string) => void;
};

/**
 * Resolved return value of {@link runInCI}
 */
export type RunResult = StandaloneRunResult | MonorepoRunResult;

/**
 * Resolved return value of {@link runInCI} in standalone mode
 */
export type StandaloneRunResult = Omit<ProjectRunResult, 'name'> & {
  mode: 'standalone';
  commentId?: number;
};

/**
 * Resolved return value of {@link runInCI} in monorepo mode
 */
export type MonorepoRunResult = {
  mode: 'monorepo';
  projects: ProjectRunResult[];
  commentId?: number;
  diffArtifact?: ArtifactData;
};

/**
 * Result of {@link runInCI} for a single project
 */
export type ProjectRunResult = {
  name: string;
  artifacts: {
    report: ArtifactData;
    diff?: ArtifactData;
  };
  newIssues?: SourceFileIssue[];
};

/**
 * Paths to output files from {@link runInCI}, for uploading as job artifact
 */
export type ArtifactData = {
  rootDir: string;
  files: string[];
};
