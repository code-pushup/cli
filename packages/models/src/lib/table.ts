import { z } from 'zod';
import { primitiveValueSchema } from './implementation/schemas';

export const tableAlignmentSchema = z.enum(['left', 'center', 'right'], {
  description: 'Cell alignment',
});
export type TableAlignment = z.infer<typeof tableAlignmentSchema>;

export const tableColumnPrimitiveSchema = tableAlignmentSchema;
export type TableColumnPrimitive = z.infer<typeof tableColumnPrimitiveSchema>;

export const tableColumnObjectSchema = z.object({
  key: z.string(),
  label: z.string().optional(),
  align: tableAlignmentSchema.optional(),
});
export type TableColumnObject = z.infer<typeof tableColumnObjectSchema>;

export const tableRowObjectSchema = z.record(primitiveValueSchema, {
  description: 'Object row',
});
export type TableRowObject = z.infer<typeof tableRowObjectSchema>;

export const tableRowPrimitiveSchema = z.array(primitiveValueSchema, {
  description: 'Primitive row',
});
export type TableRowPrimitive = z.infer<typeof tableRowPrimitiveSchema>;

const tablePrimitiveSchema = z.object(
  {
    columns: z.array(tableAlignmentSchema).optional(),
    rows: z.array(tableRowPrimitiveSchema),
  },
  { description: 'Table with primitive rows and optional alignment columns' },
);

const tableObjectSchema = z.object(
  {
    columns: z
      .union([z.array(tableAlignmentSchema), z.array(tableColumnObjectSchema)])
      .optional(),
    rows: z.array(tableRowObjectSchema),
  },
  {
    description:
      'Table with object rows and optional alignment or object columns',
  },
);

export const tableSchema = (description = 'Table information') =>
  z.union([tablePrimitiveSchema, tableObjectSchema], { description });
export type Table = z.infer<ReturnType<typeof tableSchema>>;
