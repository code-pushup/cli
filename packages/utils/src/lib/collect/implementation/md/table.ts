import { NEW_LINE } from './constants';

export type Alignment = 'l' | 'c' | 'r';
const alignString = new Map<Alignment, string>([
  ['l', ':--'],
  ['c', ':--:'],
  ['r', '--:'],
]);

/**
 * | Table Header 1  | Table Header 2 |
 * | --------------- | -------------- |
 * |  String 1       |  1             |
 * |  String 1       |  2             |
 * |  String 1       |  3             |
 */
export function table(
  data: (string | number)[][],
  align?: Alignment[],
): string {
  if (data.length === 0) {
    throw new Error("Data can't be empty");
  }
  align = align || data[0]?.map(() => 'c');
  const _data = data.map(arr => '|' + arr.join('|') + '|');
  const secondRow = '|' + align?.map(s => alignString.get(s)).join('|') + '|';
  return _data.shift() + NEW_LINE + secondRow + NEW_LINE + _data.join(NEW_LINE);
}
