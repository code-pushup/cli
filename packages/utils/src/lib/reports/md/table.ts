import { NEW_LINE } from '../constants';

export type Alignment = 'l' | 'c' | 'r';
const alignString = new Map<Alignment, string>([
  ['l', ':--'],
  ['c', ':--:'],
  ['r', '--:'],
]);

/**
 * | Table Header 1  | Table Header 2 |
 * | --------------- | -------------- |
 * |  String 1       |  1             |
 * |  String 1       |  2             |
 * |  String 1       |  3             |
 */
export function tableMd(
  data: (string | number)[][],
  align?: Alignment[],
): string {
  if (data.length === 0) {
    throw new Error("Data can't be empty");
  }
  const alignmentSetting = align ?? data[0]?.map(() => 'c');
  const tableContent = data.map(arr => `|${arr.join('|')}|`);
  const alignmentRow = `|${alignmentSetting
    ?.map(s => alignString.get(s))
    .join('|')}|`;
  return (
    tableContent[0] +
    NEW_LINE +
    alignmentRow +
    NEW_LINE +
    tableContent.slice(1).join(NEW_LINE)
  );
}

export function tableHtml(data: (string | number)[][]): string {
  if (data.length === 0) {
    throw new Error("Data can't be empty");
  }

  const tableContent = data.map((arr, index) => {
    if (index === 0) {
      const headerRow = arr.map(s => `<th>${s}</th>`).join('');
      return `<tr>${headerRow}</tr>`;
    }
    const row = arr.map(s => `<td>${s}</td>`).join('');
    return `<tr>${row}</tr>`;
  });
  return `<table>${tableContent.join('')}</table>`;
}
