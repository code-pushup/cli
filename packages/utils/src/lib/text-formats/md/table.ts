import { Table, TableAlignment } from '@code-pushup/models';
import {
  columnsToStringArray,
  getColumnAlignments,
  rowToStringArray,
} from '../../table';
import { lines, section } from './section';

const alignString = new Map<TableAlignment, string>([
  ['left', ':--'],
  ['center', ':--:'],
  ['right', '--:'],
]);

function tableRow(rows: string[]): string {
  return `|${rows.join('|')}|`;
}

/**
 * | Table Header 1  | Table Header 2 |
 * | :-------------- | -------------: |
 * |  String 1       |  1             |
 * |  String 1       |  2             |
 * |  String 1       |  3             |
 */
export function table(data: Table): string {
  if (data.rows.length === 0) {
    throw new Error("Data can't be empty");
  }

  const alignmentRow = getColumnAlignments(data).map(
    s => alignString.get(s) ?? String(alignString.get('center')),
  );

  return section(
    `${lines(
      tableRow(columnsToStringArray(data)),
      tableRow(alignmentRow),
      ...rowToStringArray(data).map(tableRow),
    )}`,
  );
}
