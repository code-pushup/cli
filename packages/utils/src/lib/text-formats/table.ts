import {
  PrimitiveValue,
  Table,
  TableAlignment,
  TableColumnObject,
  TableColumnPrimitive,
} from '@code-pushup/models';
import { capitalize } from '../transform';

export function rowToStringArray({ rows, columns = [] }: Table): string[][] {
  if (Array.isArray(rows.at(0)) && typeof columns.at(0) === 'object') {
    throw new TypeError(
      'Column can`t be object when rows are primitive values',
    );
  }

  return rows.map(row => {
    // row = ['100 ms', '200 ms']
    if (Array.isArray(row)) {
      return row.map(String);
    }

    // row = { prop1: '100 ms', prop2: '200 ms' }
    const objectRow = row;

    // columns [] || column = 'center'
    if (columns.length === 0 || typeof columns.at(0) === 'string') {
      return Object.values(objectRow).map(String);
    }

    // column = {key: 'prop1'}
    return (columns as TableColumnObject[]).map(({ key }): string =>
      String(objectRow[key]),
    );
  });
}

// Determine effective columns based on the input rows and optional columns parameter
export function columnsToStringArray({ rows, columns = [] }: Table): string[] {
  const firstRow = rows.at(0);
  const primitiveRows = Array.isArray(firstRow);

  if (typeof columns.at(0) === 'string' && !primitiveRows) {
    throw new Error('invalid union type. Caught by model parsing.');
  }

  if (columns.length === 0) {
    if (Array.isArray(firstRow)) {
      return (firstRow as unknown[]).map((_, idx) => String(idx));
    }
    return Object.keys(firstRow as object);
  }

  // columns = ['right', 'left']; row = ['100 ms', '200 ms'] => 1,2,3
  if (typeof columns.at(0) === 'string') {
    return columns.map(String);
  }

  // columns = [
  // {key: 'prop1' label: 'Property 1'},
  // {key: 'prop2'}
  // ]; row = [ {prop1: '100 ms', prop2: '200 ms'}] => Property 1, Prop2
  const cols = columns as TableColumnObject[];
  return cols.map(({ label, key }) => label ?? capitalize(key));
}

export function getColumnAlignmentForKeyAndIndex(
  targetKey: string,
  targetIdx: number,
  columns: TableColumnObject[] = [],
): TableAlignment {
  const column =
    columns.at(targetIdx) ?? columns.find(col => col.key === targetKey);
  if (typeof column === 'string') {
    return column as TableAlignment;
  } else if (typeof column === 'object') {
    return column.align ?? 'center';
  } else {
    return 'center';
  }
}

export function getColumnAlignmentForIndex(
  targetIdx: number,
  columns: TableColumnPrimitive[] = [],
): TableAlignment {
  const column = columns.at(targetIdx) as
    | TableColumnObject
    | TableAlignment
    | undefined;
  if (column == null) {
    return 'center';
  } else if (typeof column === 'string') {
    return column as TableAlignment;
  } else if (typeof column === 'object') {
    return column.align ?? 'center';
  } else {
    return 'center';
  }
}

export function getColumnAlignments({
  rows,
  columns = [],
}: Table): TableAlignment[] {
  // this is caught by the table schema in @code-pushup/models
  if (rows.at(0) == null) {
    throw new Error('first row can`t be undefined.');
  }

  if (Array.isArray(rows.at(0))) {
    const firstPrimitiveRow = rows.at(0) as PrimitiveValue[];
    return Array.from({ length: firstPrimitiveRow.length }).map((_, idx) =>
      getColumnAlignmentForIndex(idx, columns as TableColumnPrimitive[]),
    );
  }

  const firstObject = rows.at(0) as Record<string, unknown>;
  return Object.keys(firstObject).map((key, idx) =>
    getColumnAlignmentForKeyAndIndex(key, idx, columns as TableColumnObject[]),
  );
}
