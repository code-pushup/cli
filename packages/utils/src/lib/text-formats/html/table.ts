import { Table } from '@code-pushup/models';
import { NEW_LINE } from '../constants';
import { columnsToStringArray, rowToStringArray } from '../table';

function wrap(elem: string, content: string): string {
  return `<${elem}>${content}</${elem}>${NEW_LINE}`;
}

function wrapRow(content: string): string {
  const elem = 'tr';
  return `<${elem}>${NEW_LINE}${content}</${elem}>${NEW_LINE}`;
}

export function table(tableData: Table): string {
  if (tableData.rows.length === 0) {
    throw new Error("Data can't be empty");
  }

  // @TODO add formatting
  const tableHeaderCols = columnsToStringArray(tableData)
    .map(s => wrap('th', s))
    .join('');
  const tableHeaderRow = wrapRow(tableHeaderCols);
  const tableBody = rowToStringArray(tableData)
    .map(arr => {
      const columns = arr.map(s => wrap('td', s)).join('');
      return wrapRow(columns);
    })
    .join('');

  return wrap('table', `${NEW_LINE}${tableHeaderRow}${tableBody}`);
}
