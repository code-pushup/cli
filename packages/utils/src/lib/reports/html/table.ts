import { Table } from '@code-pushup/models';
import { columnsToStringArray, rowToStringArray } from '../../table';
import { NEW_LINE } from '../md/constants';

function wrap(elem: string, content: string): string {
  return `<${elem}>${content}</${elem}>${NEW_LINE}`;
}

function wrapRow(content: string): string {
  const elem = 'tr';
  return `<${elem}>${NEW_LINE}${content}</${elem}>${NEW_LINE}`;
}

export function tableHtml(table: Table): string {
  if (table.rows.length === 0) {
    throw new Error("Data can't be empty");
  }

  // @TODO add formatting
  const tableHeaderCols = columnsToStringArray(table)
    .map(s => wrap('th', s))
    .join('');
  const tableHeaderRow = wrapRow(tableHeaderCols);
  const tableBody = rowToStringArray(table)
    .map(arr => {
      const columns = arr.map(s => wrap('td', s)).join('');
      return wrapRow(columns);
    })
    .join('');

  return wrap('table', `${NEW_LINE}${tableHeaderRow}${tableBody}`);
}
