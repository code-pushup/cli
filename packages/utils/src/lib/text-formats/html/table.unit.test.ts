import { Table } from '@code-pushup/models';
import { table } from './table';

describe('tableHtml', () => {
  it('should create a complete table', () => {
    const data: Table = {
      columns: [
        { key: 'date', label: 'Date of Action', align: 'right' },
        { key: 'time', label: 'Time of Action', align: 'left' },
        { key: 'action', label: 'Action' },
      ],
      rows: [{ date: '2025.01.01', time: '00:00:00', action: 'add item' }],
    };
    expect(table(data)).toMatchSnapshot();
  });
});
