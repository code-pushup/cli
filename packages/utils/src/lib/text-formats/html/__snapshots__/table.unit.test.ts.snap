// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tableHtml > should create a complete table 1`] = `
"<table>
<tr>
<th>Date of Action</th>
<th>Time of Action</th>
<th>Action</th>
</tr>
<tr>
<td>2025.01.01</td>
<td>00:00:00</td>
<td>add item</td>
</tr>
</table>
"
`;
